lib/pure/parseopt2.nim

```diff
@@ -0,0 +1,133 @@
+    optShort ## -s
+    optLong  ## --long
+    optPos   ## positional
+  Opt* = object
+    case kind*: OptKind
+    of optShort:
+      keyShort*: char
+    of optLong:
+      keyLong*: string
+    of optPos:
+      discard
+    val*: string
+  OptError* = ref object of CatchableError
+  OptExtraneousVal* = ref object of OptError
+    case kind*: OptKind
+    of optShort:
+      keyShort*: char
+    of optLong:
+      keyLong*: string
+    of optPos:
+      discard
+  OptMissingVal* = ref object of OptError
+    case kind*: OptKind
+    of optShort:
+      keyShort*: char
+    of optLong:
+      keyLong*: string
+    of optPos:
+      discard
+
+# xxx: trivial impl. remove this once == for case objects is implemented
+proc`==`*(a, b: Opt): bool =
+  a.kind == b.kind and a.val == b.val and (
+    case a.kind:
+    of optShort:
+      a.keyShort == b.keyShort
+    of optLong:
+      a.keyLong == b.keyLong
+    of optPos:
+      true
+  )
+
+iterator opts*(argv: openArray[string], shortHasVal: openArray[char] = [], longHasVal: openArray[string] = [], sep: set[char] = {'=', ':'}): Opt =
+  var all_positional = false
+  var partial = Opt.none
+  for arg in argv:
+    if partial.isSome: # fill val of partial opt
+      var opt = partial.get
+      opt.val = arg
+      yield opt
+      partial = Opt.none
+    elif all_positional:
+      yield Opt(kind: optPos, val: arg)
+    elif arg == "-":
+      yield Opt(kind: optPos, val: arg)
+    elif arg == "--":
+      all_positional = true
+    elif arg.startsWith "--": # process long
+      # long
+      var key = ""
+      var i = 2 # skip starting --
+      block process_arg:
+        while i < arg.len:
+          let c = arg[i]
+          i.inc
+          if c in sep:
+            if key notin longHasVal:
+              raise OptExtraneousVal(kind: optLong, keyLong: key)
+            yield Opt(kind: optLong, keyLong: key, val: arg[i..^1])
+            break process_arg
+          else:
+            key &= c
+        
+        if key in longHasVal:
+          # wait for .val to be filled
+          partial = some Opt(kind: optLong, keyLong: key)
+        else:
+          yield Opt(kind: optLong, keyLong: key)
+    elif arg.startsWith "-": # process short
+      var i = 1 # skip starting -
+      while i < arg.len:
+        let c = arg[i]
+        i.inc
+        if c notin shortHasVal:
+          yield Opt(kind: optShort, keyShort: c)
+        elif i < arg.len:
+          if arg[i] in sep:
+            # skip separator (e.g. '=') in "-a=c" if it exist
+            i.inc
+          yield Opt(kind: optShort, keyShort: c, val: arg[i..^1])
+          break
+        else:
+          partial = some Opt(kind: optShort, keyShort: c)
+          break
+    else: # process positional
+      yield Opt(kind: optPos, val: arg)
+
+  # throw on missing val
+  if partial.isSome:
+    let opt = partial.get
+    case opt.kind
+    of optShort:
+      raise OptMissingVal(kind: optShort, keyShort: opt.keyShort)
```

Comment:
Counter-point, Nim is wrong. 😆 

We have `Result` (https://github.com/nim-works/nimskull/blob/devel/lib/experimental/results.nim) or this could also be handled as in-band data in the enum itself. Since this is an iterator, basically a stream of results, inband might be a better fit.